Source/WebCore/page/Navigation.cpp

```diff
@@ -379,4 +365,159 @@ void Navigation::updateForNavigation(Ref<HistoryItem>&& item, FrameLoadType fram
+
+    // Step 32:
+    if (event->wasIntercepted()) {
+        event->setInterceptionState(InterceptionState::Committed);
+
+        RefPtr fromNavigationHistoryEntry = currentEntry();
+        ASSERT(fromNavigationHistoryEntry);
+
+        // FIXME: Create finished promise
+        m_transition = NavigationTransition::create(navigationType, *fromNavigationHistoryEntry);
+
+        if (navigationType == NavigationNavigationType::Traverse)
+            m_suppressNormalScrollRestorationDuringOngoingNavigation = true;
+
+        // FIXME: Step 7: URL and history update
+    }
+
+    if (endResultIsSameDocument) {
+        // FIXME: Step 33: Wait on Handler promises
+
+        if (document->isFullyActive() && !abortController->signal().aborted()) {
+            ASSERT(m_ongoingNavigateEvent == event.ptr());
+            m_ongoingNavigateEvent = nullptr;
+
+            event->finish();
+
+            // FIXME: 6. Fire an event named navigatesuccess at navigation.
+            // FIXME: 7. If navigation's transition is not null, then resolve navigation's transition's finished promise with undefined.
+            m_transition = nullptr;
+
+            // FIXME: if (apiMethodTracker)
+            //             resolveFinishedPromise(*apiMethodTracker);
+        }
+
+        // FIXME: and the following failure steps given reason rejectionReason:
+        m_ongoingNavigateEvent = nullptr;
+    }
+
+    // FIXME: if (apiMethodTracker)
+    //            cleanupAPIMethodTracker(*apiMethodTracker);
+
+    // FIXME: Step 35 Clean up after running script
+
+    return !event->wasIntercepted();
+}
+
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#fire-a-traverse-navigate-event
+bool Navigation::dispatchTraversalNavigateEvent(Ref<HistoryItem> historyItem)
+{
+    // FIME: isCurrentDocument may not match spec
+    bool isSameDocument = historyItem->isCurrentDocument(*window()->protectedDocument());
+    // FIXME: Set destinations state
+    // FIXME: Get Entry for historyItem
+    Ref destination = NavigationDestination::create(historyItem->url(), currentEntry(), isSameDocument);
+
+    return innerDispatchNavigateEvent(NavigationNavigationType::Traverse, WTFMove(destination), { });
+}
+
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#fire-a-push/replace/reload-navigate-event
+bool Navigation::dispatchPushReplaceReloadNavigateEvent(const URL& url, NavigationNavigationType navigationType, bool isSameDocument)
+{
+    // FIXME: Set event's classic history API state to classicHistoryAPIState.
+    Ref destination = NavigationDestination::create(url, nullptr, isSameDocument);
+    return innerDispatchNavigateEvent(navigationType, WTFMove(destination), { });
+}
+
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#fire-a-download-request-navigate-event
+bool Navigation::dispatchDownloadNavigateEvent(const URL&, const String& downloadFilename)
+{
+    // FIXME
```

Comment:
When to call this is tricky, since it seems to be handled mostly in FrameLoaderClient. Maybe an option is to send the event in PolicyChecker::checkNavigationPolicy before calling startDownload.

Reply:
Yeah, the policychecker was one place I was looking.

All of these dispatches seem tricky to find the ideal spot, the general navigation specs don't seem super aligned with WebKit's structure to me.

Reply:
> Yeah, the policychecker was one place I was looking.
> 
> All of these dispatches seem tricky to find the ideal spot, the general navigation specs don't seem super aligned with WebKit's structure to me.

Correct, for example the download part has a quite seperate algorithm, but I would bet that spec was done after the implementation. I would like to work on refactoring that, but it is not crucial for Navigation API.